lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java

```diff
@@ -24,167 +31,166 @@
+        this.dataSource = requireNonNull(dataSource, "dataSource is null");
+        requireNonNull(block, "block is null");
+        requireNonNull(columnsRead, "columnsRead is null");
+        requireNonNull(columnsFiltered, "columnsFiltered is null");
 
-        TrinoIndexStore(ColumnChunkMetaData meta)
-        {
-            this.meta = meta;
-            OffsetIndex offsetIndex;
-            try {
-                offsetIndex = readOffsetIndex(meta);
-            }
-            catch (IOException e) {
-                // If the I/O issue still stands it will fail the reading later;
-                // otherwise we fail the filtering only with a missing offset index.
-                offsetIndex = null;
+        ImmutableList.Builder<ColumnIndexMetadata> columnIndexBuilder = ImmutableList.builderWithExpectedSize(columnsFiltered.size());
+        ImmutableList.Builder<ColumnIndexMetadata> offsetIndexBuilder = ImmutableList.builderWithExpectedSize(columnsRead.size());
+        for (ColumnChunkMetaData column : block.getColumns()) {
+            ColumnPath path = column.getPath();
+            if (column.getColumnIndexReference() != null && columnsFiltered.contains(path)) {
+                columnIndexBuilder.add(new ColumnIndexMetadata(
+                        column.getColumnIndexReference(),
+                        path,
+                        column.getPrimitiveType()));
             }
-            if (offsetIndex == null) {
-                throw new MissingOffsetIndexException(meta.getPath());
+            if (column.getOffsetIndexReference() != null && columnsRead.contains(path)) {
+                offsetIndexBuilder.add(new ColumnIndexMetadata(
+                        column.getOffsetIndexReference(),
+                        path,
+                        column.getPrimitiveType()));
             }
-            this.offsetIndex = offsetIndex;
         }
+        this.columnIndexReferences = columnIndexBuilder.build();
+        this.offsetIndexReferences = offsetIndexBuilder.build();
+    }
 
-        @Override
-        public ColumnIndex getColumnIndex()
-        {
-            if (!columnIndexRead) {
+    @Override
+    public ColumnIndex getColumnIndex(ColumnPath column)
+    {
+        if (columnIndexStore == null) {
+            columnIndexStore = loadIndexes(dataSource, columnIndexReferences, (buffer, columnMetadata) -> {
                 try {
-                    columnIndex = readColumnIndex(meta);
+                    return ParquetMetadataConverter.fromParquetColumnIndex(columnMetadata.getPrimitiveType(), Util.readColumnIndex(buffer.getInput()));
                 }
                 catch (IOException e) {
                     // If the I/O issue still stands it will fail the reading later;
                     // otherwise we fail the filtering only with a missing column index.
+                    LOG.warn(e, "Unable to read column index for column %s", columnMetadata.getPath());
+                    return null;
```

Comment:
There’s no actual I/O here. A failure in this context indicates the column index is corrupted, so we should fail with a proper exception.